Git clone does not work on cert providing environment

It looks like the PR about using a custom CA certificates forgot to mention that `git clone` command (executed by TG) need also use the cert. As a result, TG cannot successfully clone a project to run `renku log` on it.

The exception being thrown on TG
```
Caused by: java.lang.Exception: git clone failed with: CommandResult 128
Cloning into '/tmp/test--4883980789101307731'...
fatal: unable to access 'https://gitlab.gck.health2030.ch/renku/test.git/': 
SSL certificate problem: self-signed certificate in certificate chain```

--- a/graph-commons/src/main/scala/ch/datascience/config/certificates/Certificate.scala
+++ b/graph-commons/src/main/scala/ch/datascience/config/certificates/Certificate.scala
@@ -20,12 +20,11 @@ package ch.datascience.config.certificates
 
 import cats.MonadError
 import cats.syntax.all._
-import ch.datascience.logging.ApplicationLogger
 import ch.datascience.tinytypes.constraints.NonBlank
 import ch.datascience.tinytypes.{StringTinyType, TinyTypeFactory}
 
-private final class Certificate private (val value: String) extends AnyVal with StringTinyType
-private object Certificate extends TinyTypeFactory[Certificate](new Certificate(_)) with NonBlank {
+final class Certificate private (val value: String) extends AnyVal with StringTinyType
+object Certificate extends TinyTypeFactory[Certificate](new Certificate(_)) with NonBlank {
 
   import ch.datascience.config.ConfigLoader.find
   import com.typesafe.config.{Config, ConfigFactory}

--- a/graph-commons/src/main/scala/ch/datascience/config/certificates/CertificateLoader.scala
+++ b/graph-commons/src/main/scala/ch/datascience/config/certificates/CertificateLoader.scala
@@ -34,7 +34,7 @@ object CertificateLoader {
   )(implicit ME: MonadError[Interpretation, Throwable]): Interpretation[CertificateLoader[Interpretation]] =
     for {
       keystore <- Keystore[Interpretation]()
-    } yield new CertificateLoaderImpl(
+    } yield new CertificateLoaderImpl[Interpretation](
       keystore,
       findCertificate = () => Certificate.fromConfig[Interpretation](),
       createSslContext = (keystore: Keystore[Interpretation]) => SslContext.from(keystore),

--- a/graph-commons/src/test/scala/ch/datascience/config/certificates/CertificateLoaderSpec.scala
+++ b/graph-commons/src/test/scala/ch/datascience/config/certificates/CertificateLoaderSpec.scala
@@ -20,16 +20,17 @@ package ch.datascience.config.certificates
 
 import cats.MonadError
 import cats.syntax.all._
+import ch.datascience.generators.CommonGraphGenerators.certificates
 import ch.datascience.generators.Generators.Implicits._
-import ch.datascience.generators.Generators.{exceptions, nonBlankStrings}
+import ch.datascience.generators.Generators.exceptions
 import ch.datascience.interpreters.TestLogger
 import ch.datascience.interpreters.TestLogger.Level.{Error, Info}
-import javax.net.ssl.SSLContext
 import org.scalacheck.Gen
 import org.scalamock.scalatest.MockFactory
 import org.scalatest.matchers.should
 import org.scalatest.wordspec.AnyWordSpec
 
+import javax.net.ssl.SSLContext
 import scala.util.{Failure, Try}
 
 class CertificateLoaderSpec extends AnyWordSpec with MockFactory with should.Matchers {
@@ -138,6 +139,5 @@ class CertificateLoaderSpec extends AnyWordSpec with MockFactory with should.Mat
     )
   }
 
-  private lazy val certificates: Gen[Certificate] = nonBlankStrings().map(v => Certificate(v.value))
-  private lazy val sslContexts:  Gen[SslContext]  = Gen.uuid.map(_ => new SslContext(SSLContext.getInstance("TLS")))
+  private lazy val sslContexts: Gen[SslContext] = Gen.uuid.map(_ => new SslContext(SSLContext.getInstance("TLS")))
 }

--- a/graph-commons/src/test/scala/ch/datascience/generators/CommonGraphGenerators.scala
+++ b/graph-commons/src/test/scala/ch/datascience/generators/CommonGraphGenerators.scala
@@ -18,9 +18,10 @@
 
 package ch.datascience.generators
 
+import ch.datascience.config.certificates.Certificate
+
 import java.nio.charset.StandardCharsets.UTF_8
 import java.util.Base64
-
 import ch.datascience.config.renku
 import ch.datascience.config.sentry.SentryConfig
 import ch.datascience.config.sentry.SentryConfig.{EnvironmentName, SentryBaseUrl, ServiceName}
@@ -175,6 +176,15 @@ object CommonGraphGenerators {
       .fold(throw _, identity)
 
   implicit val fusekiBaseUrls: Gen[FusekiBaseUrl] = httpUrls() map FusekiBaseUrl.apply
+
+  implicit lazy val certificates: Gen[Certificate] =
+    nonBlankStrings()
+      .toGeneratorOfNonEmptyList(minElements = 2)
+      .map { lines =>
+        Certificate {
+          lines.toList.mkString("-----BEGIN CERTIFICATE-----\n", "\n", "\n-----END CERTIFICATE-----")
+        }
+      }
 
   implicit val jsonLDTriples: Gen[JsonLDTriples] = for {
     subject <- nonEmptyStrings()

--- a/triples-generator/src/main/scala/ch/datascience/triplesgenerator/Microservice.scala
+++ b/triples-generator/src/main/scala/ch/datascience/triplesgenerator/Microservice.scala
@@ -20,7 +20,6 @@ package ch.datascience.triplesgenerator
 
 import java.util.concurrent.ConcurrentHashMap
 import java.util.concurrent.Executors.newFixedThreadPool
-
 import cats.effect._
 import ch.datascience.config.GitLab
 import ch.datascience.config.certificates.CertificateLoader
@@ -32,6 +31,7 @@ import ch.datascience.metrics.{MetricsRegistry, RoutesMetrics}
 import ch.datascience.microservices.IOMicroservice
 import ch.datascience.rdfstore.SparqlQueryTimeRecorder
 import ch.datascience.triplesgenerator.config.TriplesGeneration
+import ch.datascience.triplesgenerator.config.certificates.GitCertificateInstaller
 import ch.datascience.triplesgenerator.eventprocessing._
 import ch.datascience.triplesgenerator.init._
 import ch.datascience.triplesgenerator.reprovisioning.{IOReProvisioning, ReProvisioning, ReProvisioningStatus}
@@ -57,6 +57,7 @@ object Microservice extends IOMicroservice {
   override def run(args: List[String]): IO[ExitCode] =
     for {
       certificateLoader        <- CertificateLoader[IO](ApplicationLogger)
+      gitCertificateInstaller  <- GitCertificateInstaller[IO](ApplicationLogger)
       sentryInitializer        <- SentryInitializer[IO]()
       fusekiDatasetInitializer <- IOFusekiDatasetInitializer()
       subscriber               <- Subscriber(ApplicationLogger)
@@ -80,6 +81,7 @@ object Microservice extends IOMicroservice {
       exitCode <- microserviceRoutes.use { routes =>
                     new MicroserviceRunner(
                       certificateLoader,
+                      gitCertificateInstaller,
                       sentryInitializer,
                       fusekiDatasetInitializer,
                       subscriber,
@@ -93,6 +95,7 @@ object Microservice extends IOMicroservice {
 
 private class MicroserviceRunner(
     certificateLoader:        CertificateLoader[IO],
+    gitCertificateInstaller:  GitCertificateInstaller[IO],
     sentryInitializer:        SentryInitializer[IO],
     datasetInitializer:       FusekiDatasetInitializer[IO],
     subscriber:               Subscriber[IO],
@@ -103,6 +106,7 @@ private class MicroserviceRunner(
 
   def run(): IO[ExitCode] = for {
     _        <- certificateLoader.run()
+    _        <- gitCertificateInstaller.run()
     _        <- sentryInitializer.run()
     _        <- datasetInitializer.run()
     _        <- subscriber.run().start.map(gatherCancelToken)

--- a/triples-generator/src/main/scala/ch/datascience/triplesgenerator/config/certificates/CertificateSaver.scala
+++ b/triples-generator/src/main/scala/ch/datascience/triplesgenerator/config/certificates/CertificateSaver.scala
@@ -1,0 +1,55 @@
+/*
+ * Copyright 2020 Swiss Data Science Center (SDSC)
+ * A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+ * Eidgenössische Technische Hochschule Zürich (ETHZ).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ch.datascience.triplesgenerator.config.certificates
+
+import cats.MonadError
+import ch.datascience.config.certificates.Certificate
+
+import java.io.{File, PrintWriter}
+import java.nio.file.{Files, Path, Paths}
+
+private trait CertificateSaver[Interpretation[_]] {
+  def save(certificate: Certificate): Interpretation[Path]
+}
+
+private object CertificateSaver {
+  def apply[Interpretation[_]]()(implicit ME: MonadError[Interpretation, Throwable]): CertificateSaver[Interpretation] =
+    new CertificateSaverImpl[Interpretation]()
+}
+
+private class CertificateSaverImpl[Interpretation[_]](implicit ME: MonadError[Interpretation, Throwable])
+    extends CertificateSaver[Interpretation] {
+
+  private val certPathDirectory: Path = Paths.get(s"${System.getProperty("user.home")}${File.separator}/git-certs")
+  private val certPath:          Path = certPathDirectory.resolve("cert.pem")
+
+  override def save(certificate: Certificate): Interpretation[Path] = ME.catchNonFatal {
+
+    Files.createDirectory(certPathDirectory)
+
+    val certFile = Files.createFile(certPath).toFile
+
+    val pw = new PrintWriter(certFile)
+
+    try pw.write(certificate.toString)
+    finally pw.close()
+
+    certPath
+  }
+}

--- a/triples-generator/src/main/scala/ch/datascience/triplesgenerator/config/certificates/GitCertificateInstaller.scala
+++ b/triples-generator/src/main/scala/ch/datascience/triplesgenerator/config/certificates/GitCertificateInstaller.scala
@@ -1,0 +1,70 @@
+/*
+ * Copyright 2020 Swiss Data Science Center (SDSC)
+ * A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+ * Eidgenössische Technische Hochschule Zürich (ETHZ).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ch.datascience.triplesgenerator.config.certificates
+
+import cats.MonadError
+import ch.datascience.config.certificates.Certificate
+import io.chrisdavenport.log4cats.Logger
+
+import scala.util.control.NonFatal
+
+trait GitCertificateInstaller[Interpretation[_]] {
+  def run(): Interpretation[Unit]
+}
+
+object GitCertificateInstaller {
+  def apply[Interpretation[_]](
+      logger:    Logger[Interpretation]
+  )(implicit ME: MonadError[Interpretation, Throwable]): Interpretation[GitCertificateInstaller[Interpretation]] =
+    ME.catchNonFatal {
+      new GitCertificateInstallerImpl[Interpretation](
+        () => Certificate.fromConfig[Interpretation](),
+        CertificateSaver(),
+        GitConfigModifier(),
+        logger
+      )
+    }
+}
+
+class GitCertificateInstallerImpl[Interpretation[_]](
+    findCertificate:   () => Interpretation[Option[Certificate]],
+    certificateSaver:  CertificateSaver[Interpretation],
+    gitConfigModifier: GitConfigModifier[Interpretation],
+    logger:            Logger[Interpretation]
+)(implicit ME:         MonadError[Interpretation, Throwable])
+    extends GitCertificateInstaller[Interpretation] {
+
+  import cats.syntax.all._
+
+  def run(): Interpretation[Unit] = findCertificate() flatMap {
+    case None => ().pure[Interpretation]
+    case Some(certificate) =>
+      for {
+        certPath <- certificateSaver.save(certificate)
+        _        <- gitConfigModifier.makeGitTrust(certPath)
+        _        <- logger info "Certificate installed for Git"
+      } yield ()
+  } recoverWith logMessage
+
+  private lazy val logMessage: PartialFunction[Throwable, Interpretation[Unit]] = { case NonFatal(exception) =>
+    logger
+      .error(exception)("Certificate installation for Git failed")
+      .flatMap(_ => exception.raiseError[Interpretation, Unit])
+  }
+}

--- a/triples-generator/src/main/scala/ch/datascience/triplesgenerator/config/certificates/GitConfigModifier.scala
+++ b/triples-generator/src/main/scala/ch/datascience/triplesgenerator/config/certificates/GitConfigModifier.scala
@@ -1,0 +1,45 @@
+/*
+ * Copyright 2020 Swiss Data Science Center (SDSC)
+ * A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+ * Eidgenössische Technische Hochschule Zürich (ETHZ).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ch.datascience.triplesgenerator.config.certificates
+
+import cats.MonadError
+
+import java.nio.file.Path
+
+private trait GitConfigModifier[Interpretation[_]] {
+  def makeGitTrust(certPath: Path): Interpretation[Unit]
+}
+
+private object GitConfigModifier {
+  def apply[Interpretation[_]]()(implicit
+      ME: MonadError[Interpretation, Throwable]
+  ): GitConfigModifier[Interpretation] =
+    new GitConfigModifierImpl[Interpretation]()
+}
+
+private class GitConfigModifierImpl[Interpretation[_]]()(implicit
+    ME: MonadError[Interpretation, Throwable]
+) extends GitConfigModifier[Interpretation] {
+  import ammonite.ops.{%%, home}
+  import cats.syntax.all._
+
+  override def makeGitTrust(certPath: Path): Interpretation[Unit] = ME.catchNonFatal {
+    %%("git", "config", "--global", "http.sslCAInfo", certPath.toAbsolutePath.toString)(home)
+  }.void
+}

--- a/triples-generator/src/main/scala/ch/datascience/triplesgenerator/eventprocessing/triplesgeneration/renkulog/Commands.scala
+++ b/triples-generator/src/main/scala/ch/datascience/triplesgenerator/eventprocessing/triplesgeneration/renkulog/Commands.scala
@@ -29,8 +29,7 @@ import ch.datascience.graph.model.projects
 import ch.datascience.http.client.AccessToken
 import ch.datascience.http.client.AccessToken.{OAuthAccessToken, PersonalAccessToken}
 import ch.datascience.rdfstore.JsonLDTriples
-import ch.datascience.tinytypes.constraints.{NonBlank, RelativePath, RelativePathOps}
-import ch.datascience.tinytypes.{RelativePathTinyType, TinyType, TinyTypeFactory}
+import ch.datascience.tinytypes.{TinyType, TinyTypeFactory}
 import ch.datascience.triplesgenerator.eventprocessing.CommitEvent
 import ch.datascience.triplesgenerator.eventprocessing.CommitEvent._
 import ch.datascience.triplesgenerator.eventprocessing.CommitEventProcessor.ProcessingRecoverableError

--- a/triples-generator/src/test/scala/ch/datascience/triplesgenerator/MicroserviceRunnerSpec.scala
+++ b/triples-generator/src/test/scala/ch/datascience/triplesgenerator/MicroserviceRunnerSpec.scala
@@ -19,7 +19,6 @@
 package ch.datascience.triplesgenerator
 
 import java.util.concurrent.ConcurrentHashMap
-
 import cats.effect._
 import ch.datascience.config.certificates.CertificateLoader
 import ch.datascience.generators.Generators.Implicits._
@@ -27,6 +26,7 @@ import ch.datascience.generators.Generators._
 import ch.datascience.http.server.IOHttpServer
 import ch.datascience.interpreters.IOSentryInitializer
 import ch.datascience.testtools.MockedRunnableCollaborators
+import ch.datascience.triplesgenerator.config.certificates.GitCertificateInstaller
 import ch.datascience.triplesgenerator.init.IOFusekiDatasetInitializer
 import ch.datascience.triplesgenerator.reprovisioning.ReProvisioning
 import ch.datascience.triplesgenerator.subscriptions.Subscriber
@@ -49,6 +49,7 @@ class MicroserviceRunnerSpec
       "and subscription, re-provisioning and the http server start up" in new TestCase {
 
         given(certificateLoader).succeeds(returning = ())
+        given(gitCertificateInstaller).succeeds(returning = ())
         given(sentryInitializer).succeeds(returning = ())
         given(datasetInitializer).succeeds(returning = ())
         given(subscriber).succeeds(returning = ())
@@ -68,9 +69,21 @@ class MicroserviceRunnerSpec
       } shouldBe exception
     }
 
+    "fail if installing Certificate for Git fails" in new TestCase {
+
+      given(certificateLoader).succeeds(returning = ())
+      val exception = exceptions.generateOne
+      given(gitCertificateInstaller).fails(becauseOf = exception)
+
+      intercept[Exception] {
+        runner.run().unsafeRunSync()
+      } shouldBe exception
+    }
+
     "fail if Sentry initialization fails" in new TestCase {
 
       given(certificateLoader).succeeds(returning = ())
+      given(gitCertificateInstaller).succeeds(returning = ())
       val exception = exceptions.generateOne
       given(sentryInitializer).fails(becauseOf = exception)
 
@@ -82,6 +95,7 @@ class MicroserviceRunnerSpec
     "fail if RDF dataset verification fails" in new TestCase {
 
       given(certificateLoader).succeeds(returning = ())
+      given(gitCertificateInstaller).succeeds(returning = ())
       given(sentryInitializer).succeeds(returning = ())
       val exception = exceptions.generateOne
       given(datasetInitializer).fails(becauseOf = exception)
@@ -94,6 +108,7 @@ class MicroserviceRunnerSpec
     "fail if starting the Http Server fails" in new TestCase {
 
       given(certificateLoader).succeeds(returning = ())
+      given(gitCertificateInstaller).succeeds(returning = ())
       given(sentryInitializer).succeeds(returning = ())
       given(datasetInitializer).succeeds(returning = ())
       given(subscriber).succeeds(returning = ())
@@ -109,6 +124,7 @@ class MicroserviceRunnerSpec
     "return Success ExitCode even if running Subscriber fails" in new TestCase {
 
       given(certificateLoader).succeeds(returning = ())
+      given(gitCertificateInstaller).succeeds(returning = ())
       given(sentryInitializer).succeeds(returning = ())
       given(datasetInitializer).succeeds(returning = ())
       given(subscriber).fails(becauseOf = exceptions.generateOne)
@@ -121,6 +137,7 @@ class MicroserviceRunnerSpec
     "return Success ExitCode even if starting re-provisioning process fails" in new TestCase {
 
       given(certificateLoader).succeeds(returning = ())
+      given(gitCertificateInstaller).succeeds(returning = ())
       given(sentryInitializer).succeeds(returning = ())
       given(datasetInitializer).succeeds(returning = ())
       given(subscriber).succeeds(returning = ())
@@ -132,14 +149,16 @@ class MicroserviceRunnerSpec
   }
 
   private trait TestCase {
-    val certificateLoader  = mock[CertificateLoader[IO]]
-    val sentryInitializer  = mock[IOSentryInitializer]
-    val datasetInitializer = mock[IOFusekiDatasetInitializer]
-    val subscriber         = mock[Subscriber[IO]]
-    val reProvisioning     = mock[ReProvisioning[IO]]
-    val httpServer         = mock[IOHttpServer]
+    val certificateLoader       = mock[CertificateLoader[IO]]
+    val gitCertificateInstaller = mock[GitCertificateInstaller[IO]]
+    val sentryInitializer       = mock[IOSentryInitializer]
+    val datasetInitializer      = mock[IOFusekiDatasetInitializer]
+    val subscriber              = mock[Subscriber[IO]]
+    val reProvisioning          = mock[ReProvisioning[IO]]
+    val httpServer              = mock[IOHttpServer]
     val runner = new MicroserviceRunner(
       certificateLoader,
+      gitCertificateInstaller,
       sentryInitializer,
       datasetInitializer,
       subscriber,

--- a/triples-generator/src/test/scala/ch/datascience/triplesgenerator/config/certificates/CertificateSaverSpec.scala
+++ b/triples-generator/src/test/scala/ch/datascience/triplesgenerator/config/certificates/CertificateSaverSpec.scala
@@ -1,0 +1,53 @@
+/*
+ * Copyright 2020 Swiss Data Science Center (SDSC)
+ * A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+ * Eidgenössische Technische Hochschule Zürich (ETHZ).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ch.datascience.triplesgenerator.config.certificates
+
+import ch.datascience.generators.CommonGraphGenerators.certificates
+import ch.datascience.generators.Generators.Implicits._
+import org.scalatest.matchers.should
+import org.scalatest.wordspec.AnyWordSpec
+
+import java.nio.file.Files
+import scala.io.Source
+import scala.util.{Success, Try}
+
+class CertificateSaverSpec extends AnyWordSpec with should.Matchers {
+
+  "save" should {
+
+    "save the given certificate into a file" in new TestCase {
+      val certificate = certificates.generateOne
+
+      val Success(path) = certSaver.save(certificate)
+
+      try {
+        val source = Source.fromFile(path.toAbsolutePath.toString)
+        source.getLines().toList shouldBe certificate.toString.split("\n").toList
+        source.close()
+      } finally {
+        Files.deleteIfExists(path) && Files.deleteIfExists(path.getParent)
+        ()
+      }
+    }
+  }
+
+  private trait TestCase {
+    val certSaver = new CertificateSaverImpl[Try]()
+  }
+}

--- a/triples-generator/src/test/scala/ch/datascience/triplesgenerator/config/certificates/GitCertificateInstallerSpec.scala
+++ b/triples-generator/src/test/scala/ch/datascience/triplesgenerator/config/certificates/GitCertificateInstallerSpec.scala
@@ -1,0 +1,127 @@
+/*
+ * Copyright 2020 Swiss Data Science Center (SDSC)
+ * A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+ * Eidgenössische Technische Hochschule Zürich (ETHZ).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ch.datascience.triplesgenerator.config.certificates
+
+import cats.Applicative
+import cats.syntax.all._
+import ch.datascience.config.certificates.Certificate
+import ch.datascience.generators.CommonGraphGenerators.certificates
+import ch.datascience.generators.Generators.Implicits._
+import ch.datascience.generators.Generators._
+import ch.datascience.interpreters.TestLogger
+import ch.datascience.interpreters.TestLogger.Level.{Error, Info}
+import org.scalamock.scalatest.MockFactory
+import org.scalatest.matchers.should
+import org.scalatest.wordspec.AnyWordSpec
+
+import java.nio.file.{Path, Paths}
+import scala.util.Try
+
+class GitCertificateInstallerSpec extends AnyWordSpec with should.Matchers with MockFactory {
+
+  "run" should {
+
+    "save the certificate into a file and configure git to trust it" in new TestCase {
+
+      val certificate = certificates.generateOne
+      findCertificate.expects().returning(Some(certificate).pure[Try])
+
+      val path = Paths.get(relativePaths().generateOne)
+      (certificateSaver
+        .save(_: Certificate))
+        .expects(certificate)
+        .returning(path.pure[Try])
+
+      (gitConfigModifier
+        .makeGitTrust(_: Path))
+        .expects(path)
+        .returning(Applicative[Try].unit)
+
+      certInstaller.run() shouldBe Applicative[Try].unit
+
+      logger.loggedOnly(Info("Certificate installed for Git"))
+    }
+
+    "do nothing if there's no certificate found" in new TestCase {
+
+      findCertificate.expects().returning(None.pure[Try])
+
+      certInstaller.run() shouldBe Applicative[Try].unit
+    }
+
+    "fail if finding the certificate fails" in new TestCase {
+
+      val exception = exceptions.generateOne
+      findCertificate.expects().returning(exception.raiseError[Try, Option[Certificate]])
+
+      certInstaller.run() shouldBe exception.raiseError[Try, Unit]
+
+      logger.loggedOnly(Error("Certificate installation for Git failed", exception))
+    }
+
+    "fail if saving the certificate fails" in new TestCase {
+
+      val certificate = certificates.generateOne
+      findCertificate.expects().returning(Some(certificate).pure[Try])
+
+      val exception = exceptions.generateOne
+      (certificateSaver
+        .save(_: Certificate))
+        .expects(certificate)
+        .returning(exception.raiseError[Try, Path])
+
+      certInstaller.run() shouldBe exception.raiseError[Try, Unit]
+
+      logger.loggedOnly(Error("Certificate installation for Git failed", exception))
+    }
+
+    "fail if changing Git config fails" in new TestCase {
+
+      val certificate = certificates.generateOne
+      findCertificate.expects().returning(Some(certificate).pure[Try])
+
+      val path = Paths.get(relativePaths().generateOne)
+      (certificateSaver
+        .save(_: Certificate))
+        .expects(certificate)
+        .returning(path.pure[Try])
+
+      val exception = exceptions.generateOne
+      (gitConfigModifier
+        .makeGitTrust(_: Path))
+        .expects(path)
+        .returning(exception.raiseError[Try, Unit])
+
+      certInstaller.run() shouldBe exception.raiseError[Try, Unit]
+
+      logger.loggedOnly(Error("Certificate installation for Git failed", exception))
+    }
+  }
+
+  private trait TestCase {
+
+    val findCertificate   = mockFunction[Try[Option[Certificate]]]
+    val certificateSaver  = mock[CertificateSaver[Try]]
+    val gitConfigModifier = mock[GitConfigModifier[Try]]
+    val logger            = TestLogger[Try]()
+
+    val certInstaller =
+      new GitCertificateInstallerImpl[Try](findCertificate, certificateSaver, gitConfigModifier, logger)
+  }
+}

--- a/version.sbt
+++ b/version.sbt
@@ -1,1 +1,1 @@
-version in ThisBuild := "1.20.3"
+version in ThisBuild := "1.20.4"

